### PR TITLE
[AIR] Skip checkpoint cast if checkpoint is same type

### PR DESCRIPTION
**Pull request by bveeramani**
Signed-off-by: Balaji Veeramani <balaji@anyscale.com>

## Why are these changes needed?

Checkpoints attributes are reset when passed to `TensorflowPredictor.from_checkpoint`. 

`TensorflowPredictor.from_checkpoint` calls `TensorflowCheckpoint.from_checkpoint`, which creates a new checkpoint with default object attributes -- even if the passed-in checkpoint is already a `TensorflowCheckpoint`.

These changes are needed to merge https://github.com/ray-project/ray/pull/28474.

## Related issue number

See https://github.com/ray-project/ray/pull/28474 and see https://github.com/ray-project/ray/pull/26777

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/air/checkpoint.py
Comment (jiaodong) [inline python/ray/air/checkpoint.py:470]: this behavior makes sense to me but I need another pair of eyes with full checkpoint context. @krfricke and @xwjiang2010 ?

**Comment by matthewdeng** [comment]
Can you explain why it's created with default attributes? Is that not a bug?

**Comment by bveeramani** [comment]
> Can you explain why it's created with default attributes? Is that not a bug?

It's sort-of a bug. More so that the behavior of `Checkpoint.from_checkpoint` is poorly defined.

`TensorflowPredictor.from_checkpoint` casts the input checkpoint to a `TensorflowCheckpoint` by calling `TensorflowCheckpoint.from_checkpoint`.

```
checkpoint = TensorflowCheckpoint.from_checkpoint(checkpoint)
model_weights = checkpoint.get_model_weights()
preprocessor = checkpoint.get_preprocessor()
```

Attributes are restored at the `from_dict` / `from_directory` layer of abstraction. Because `TensorflowCheckpoint.from_checkpoint` creates a new checkpoint with the constructor as opposed to `from_dict` / `from_directory`, the attributes aren't restored.

```
T = TypeVar(bound=Checkpoint)

def from_checkpoint(cls: T, other: Checkpoint) -> T:
    return cls(
        local_path=other._local_path,
        data_dict=other._data_dict,
        uri=other._uri,
        obj_ref=other._obj_ref,
    )
```

An alternative implementation would be to call `from_*` methods in `from_checkpoint`.

```
def from_checkpoint(cls: T, other: Checkpoint) -> T:
    if other._data_dict:
        return cls.from_dict(other._data_dict)
    ...
```

The problem with this approach is that we'd also restore the checkpoint type. 

```
>>> checkpoint: TorchCheckpoint = ...
>>> new_checkpoint = Checkpoint.from_checkpoint(checkpoint)
>>> type(new_checkpoint). # Arguably should be `Checkpoint`
TorchCheckpoint   
```

In any case, `from_checkpoint` is a short-term hack. We should handle it soon, either by removing it or by making it an implementation detail

**Comment by matthewdeng** [comment]
Ah, thanks for the explanation.

> In any case, from_checkpoint is a short-term hack. We should handle it soon, either by removing it or by making it an implementation detail

Can this be documented either as a Github issue and/or in the code as a TODO? It would be great to have this context captured so we (a) know how to design a better long-term fix, and (b) can reasonably evaluate short-term fixes (e.g. this PR).

File: python/ray/air/checkpoint.py
Comment (krfricke) [reply python/ray/air/checkpoint.py:470]: Main question here is if we should be less strict by using `isinstance(other, type)`, which will also return True for inherited classes.

**Review by krfricke — APPROVED** [review]
LGTM, quick question before merge

**Review by amogkam — APPROVED** [review]

**Review by amogkam — CHANGES_REQUESTED** [review]
Actually thinking about it some more @bveeramani, do we even need to use `from_checkpoint` in predictors?

The predictors can used the passed in checkpoints directly without needing to call `from_checkpoint`.

**Comment by bveeramani** [comment]
> Main question here is if we should be less strict by using `isinstance(other, cls)`, which will also return True for inherited classes.

@krfricke decision was somewhat arbitrary, but the motivation is to make `cls.from_checkpoint` always return a `cls`.

Currently, if you do

```
>>> torch_checkpoint = TorchCheckpoint.from_state_dict(...)
>>> checkpoint = Checkpoint.from_checkpoint(torch_checkpoint)
```

you'd get back a `Checkpoint`

```
>>> type(checkpoint)
<class 'Checkpoint'>
```

but if we did `isinstance(other, cls)`, you'd call `Checkpoint.from_checkpoint` but get a `TorchCheckpoint` back. 

```
>>> type(checkpoint>
<class 'TorchCheckpoint'>
```

**Review by amogkam — APPROVED** [review]
> Actually thinking about it some more @bveeramani, do we even need to use from_checkpoint in predictors?

Seems like this is still necessary until https://github.com/ray-project/ray/issues/28910 is fixed.

This PR looks good as a short term fix

**Comment by bveeramani** [comment]
> Actually thinking about it some more @bveeramani, do we even need to use `from_checkpoint` in predictors?
> 
> The predictors can used the passed in checkpoints directly without needing to call `from_checkpoint`.

If we want to support `TorchPredictor.from_checkpoint(checkpoint: Checkpoint)` (as opposed to `TorchPredictor.from_checkpoint(checkpoint: TorchCheckpoint)`), then we need to call `Checkpoint.from_checkpoint` in predictors.

But yeah, in any case, we should try to unblock #28474

**Pull request merged**
